Support SPI driver for Renesas RA6, RA4, RA2 devices

thaoluonguw opened the pull request:
Add support for SPI driver for RA6, RA4, RA2 devices

- RA6: EK_RA6E2, EK_RA6M1, EK_RA6M2, EK_RA6M3, EK_RA6M4, EK_RA6M5, FPB_RA6E1, FPB_RA6E2

- RA4: EK_RA4W1, EK_RA4M2, EK_RA4M3, EK_RA4E2

- RA2: EK_RA2A1

soburi commented:
Does this work with RA2, 4, 8?

thaoluonguw commented:
Hello @soburi, 
> Does this work with RA2, 4, 8?

It expects to work with RA2 and RA4. We are updating device tree and verifying for RA4 and RA2. They will be update later or in another PR.

For RA8, it already supported. It uses sci_b (spi_b_renesas_ra8.c).

soburi commented:
> Hello @soburi,
> 
> > Does this work with RA2, 4, 8?
> 
> It expects to work with RA2 and RA4. We are updating device tree and verifying for RA4 and RA2. They will be update later or in another PR.
> 
> For RA8, it already supported. It uses sci_b (spi_b_renesas_ra8.c).

Understood. I was worried that the implementation would need to be tweaked for each model.

tbursztyka commented on drivers/spi/spi_renesas_ra.c:
```diff
@@ -0,0 +1,826 @@
+	case SPI_EVENT_ERR_FRAMING:       /* Framing error */
+	case SPI_EVENT_ERR_MODE_UNDERRUN: /* Underrun error */
+		spi_context_cs_control(&data->ctx, false);
+		spi_context_complete(&data->ctx, dev, 0);
```
this path seems to be the one for errors, but you still return 0 as code? (it should be -EIO)

thaoluonguw replied:
@tbursztyka: Thank you for your feedback, we updated it. Could you please re-check again?

tbursztyka reviewed (changes requested):
All seems quite fine to me (but for the tiny comment I made on the code part)

Aside of this, could you split your first commit in 3:
- one for the driver (its dts bindings and the actual driver)
- one for the boards 
- and one for the tests

khoa-nguyen-18 committed:
drivers: spi: Add support SPI driver for Renesas RA6, RA4, RA2

- Add SPI driver support for RA
- RA2A1 not support slave select keeping level so disable it
in Kconfig

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
tests: drivers: spi: Support test spi_loopback for RA6, RA4, RA2

Add overlay and add config to support test spi_loopback for:
ek_ra6m1, ek_ra6m2, ek_ra6m3, ek_ra6m4, ek_ra6m5, ek_ra6e2,
fpb_ra6e1, fpb_ra6e2, ek_ra4w1, ek_ra4m2, ek_ra4m3, ek_ra4e2,
ek_ra2a1.

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

tbursztyka commented:
Can you rebase and push again? CI is failing but not from your commits

ydamigos commented on drivers/spi/spi_renesas_ra.c:
```diff
@@ -0,0 +1,826 @@
+		.fsp_config =                                                                      \
+			{                                                                          \
+				.channel = DT_INST_PROP(index, channel),                           \
+				.rxi_ipl = DT_INST_IRQ_BY_NAME(index, rxi, priority),              \
```
IRQ_CONNECT() may alter the priority passed. Please check z_arm_irq_priority_set() function.
Subsequent calls to  `R_SPI_Open()` will change this priority.
It will be good to update all `*_ipl` after  IRQ_CONNECT().

ydamigos commented on drivers/spi/Kconfig.renesas_ra:
```diff
@@ -0,0 +1,33 @@
+	bool "Renesas RA SPI"
+	default y
+	depends on DT_HAS_RENESAS_RA_SPI_ENABLED
+	select USE_RA_FSP_SPI
```
Should we add	`select PINCTRL`?

khoa-nguyen-18 replied:
Thank you for your feedback, we have added it, Could you please re-check again?

Commit:
dts: arm: renesas: Add SPI support for RA6, RA4, RA2

Add device node support SPI driver for ra6-cm4, ra6-cm33,
ra4-cm4, ra4-cm33, ra2xx MCU

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 committed:
boards: renesas: Add SPI support for Renesas RA6, RA4, RA2

- Add SPI support for ek_ra6m1, ek_ra6m2, ek_ra6m3, ek_ra6m4,
ek_ra6m5, ek_ra6e2, fpb_ra6e1, fpb_ra6e2, ek_ra4e2, ek_ra4m2,
ek_ra4m3, ek_ra4w1, ek_ra2a1

- Add SPI support doc for these board

Signed-off-by: Tri Nguyen <tri.nguyen.wj@bp.renesas.com>
Signed-off-by: Thao Luong <thao.luong.uw@renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khoa-nguyen-18 commented:
Add and merge SPI support for RA4, RA2 commits into SPI support for RA6 commits

khoa-nguyen-18 commented:
just rebase main to solve conflict

khoa-nguyen-18 commented:
rebase main to solve conflicts

stmofek commented:
Hallo, 
since it’s been open for a while, I wanted to check if there are any outstanding issues or additional information needed to move it forward.
If there are any changes, tests, or documentation required, please let me know. I’m happy to address them promptly to help get the MR merged.
Best regards

tbursztyka commented:
@stmofek every time a PR is updated, it looses its previous approvals.
